client/packages/system/src/ContactTrace/DetailView/DetailView.tsx

```diff
@@ -0,0 +1,270 @@
+      },
+      isLoading: false,
+    });
+  }, [registries, isLoadingPatient, isLoadingRegistry]);
+
+  return result;
+};
+
+export type DetailViewProps = {
+  createPatientId: string | null;
+  createType: string | null;
+};
+
+export const DetailView: FC<DetailViewProps> = ({
+  createType,
+  createPatientId,
+}) => {
+  const t = useTranslation('dispensary');
+  const navigate = useNavigate();
+  const { setSuffix } = useBreadcrumbs([AppRoute.ContactTrace]);
+  const dateFormat = useFormatDateTime();
+  const { getLocalisedFullName } = useIntlUtils();
+  const id = useContactTraces.utils.idFromUrl();
+  const { data: contactData, isLoading } = useContactTraceData(
+    id,
+    createType,
+    createPatientId
+  );
+  // Used when creating a new contact trace:
+  const [newContactTraceId, setNewContactTraceId] = useState(id);
+
+  const handleSave = useContactTraces.document.upsertDocument(
+    contactData?.patient?.id ?? '',
+    contactData?.type ?? '',
+    contactTrace => {
+      if (contactTrace.id !== id) setNewContactTraceId(contactTrace.id);
+    }
+  );
+
+  const {
+    JsonForm,
+    data,
+    setData,
+    saveData,
+    isDirty,
+    validationError,
+    revert,
+  } = useJsonForms(
+    contactData?.documentName,
+    contactData?.patient?.id,
+    {
+      handleSave,
+    },
+    createType && contactData?.schema
+      ? {
+          data: contactData?.contactTrace,
+          schema: contactData?.schema,
+          isCreating: newContactTraceId === id,
+        }
+      : undefined
+  );
+
+  // When a contact trace id changes (contact trace has been created), wait till the isDirty flag
+  // is cleared and then navigate to the correct url.
+  useEffect(() => {
+    if (!isDirty && newContactTraceId !== id) {
+      navigate(
+        RouteBuilder.create(AppRoute.Dispensary)
+          .addPart(AppRoute.ContactTrace)
+          .addPart(newContactTraceId)
+          .build()
+      );
+    }
+  }, [isDirty, newContactTraceId]);
+
+  const updateContactTrace = useDebounceCallback(
+    (patch: Partial<ContactTrace>) =>
+      setData({
+        ...(typeof data === 'object' ? data : {}),
+        ...patch,
+      }),
+    [data, setData]
+  );
+
+  useEffect(() => {
+    if (contactData) {
+      setSuffix(
+        <span key="patient-contact-trace">
+          <Breadcrumb
+            to={RouteBuilder.create(AppRoute.Dispensary)
+              .addPart(AppRoute.Patients)
+              .addPart(contactData.patient.id)
+              .addQuery({ tab: PatientTabValue.ContactTraces })
+              .build()}
+          >
+            {getLocalisedFullName(
+              contactData.patient?.firstName,
+              contactData.patient?.lastName
+            )}
+          </Breadcrumb>
+          <span>{` / ${contactData.programName} - ${dateFormat.localisedDate(
+            contactData.contactTrace.datetime
+          )}`}</span>
+        </span>
+      );
+    }
+  }, [contactData]);
```

Comment:
🤔  the hard coded ` / ` is troubling me a little.  I wonder if a utility method in the breadcrumb files would be useful?

with 
```
  const { appendBreadcrumb, setSuffix } = useBreadcrumbs([
    AppRoute.ContactTrace,
  ]);
```

and the update in useBreadcrumbs

```
type BreadcrumbState = {
  appendBreadcrumb: (element: ReactElement) => void;
  setSuffix: (suffix?: string | ReactElement) => void;
  suffix?: (string | ReactElement)[];
  setUrlParts: (urlParts: UrlPart[]) => void;
  urlParts: UrlPart[];
};

const useBreadcrumbState = create<BreadcrumbState>(set => ({
  setSuffix: suffix =>
    set(state => ({ ...state, suffix: suffix ? [suffix] : undefined })),
  appendBreadcrumb: element =>
    set(state => {
      const suffix = state.suffix ? [...state.suffix, element] : [element];
      return {
        ...state,
        suffix,
      };
    }),
  suffix: undefined,
  setUrlParts: (urlParts: UrlPart[]) => set(state => ({ ...state, urlParts })),
  urlParts: [],
}));
```

you can
```
  useEffect(() => {
    if (contactData) {
      const LinkToPatient = (
        <Breadcrumb
          to={RouteBuilder.create(AppRoute.Dispensary)
            .addPart(AppRoute.Patients)
            .addPart(contactData.patient.id)
            .addQuery({ tab: PatientTabValue.ContactTraces })
            .build()}
        >
          {getLocalisedFullName(
            contactData.patient?.firstName,
            contactData.patient?.lastName
          )}
        </Breadcrumb>
      );
      const ProgramDetails = (
        <span>{`${contactData.programName} - ${dateFormat.localisedDate(
          contactData.contactTrace.datetime
        )}`}</span>
      );

      setSuffix(LinkToPatient);
      appendBreadcrumb(ProgramDetails);
    }
  }, [contactData]);
```

Reply:
what do you mean with hard coded `/`? You mean that last program details parts looks like its another step?

Can't we just move the ProgramDetails out of the breadcrumb? e.g. right align it?

`appendBreadcrumb` is like `setPostfix` right? you can only append one more part?

Reply:
> what do you mean with hard coded /? You mean that the last program details parts looks like its another step?

No, just that there is a `/` as a string hard coded in this sort of line

```
<span>{` / ${contactData.programName} - ${dateFormat.localisedDate(
            contactData.contactTrace.datetime
          )}`}</span>
```

and that feel inelegant. The rest of the breadcrumb is an array of items, and the breadcrumb component decides how to format and display the parts. Now, with the code above, we are appending a big string and essentially overriding the function of the breadcrumb display component.

> Can't we just move the ProgramDetails out of the breadcrumb? e.g. right align it?

Could do, I guess. The original intention of the `setSuffix` was to have control over the formatting of the last portion of the breadcrumb, and I think that's still valid here. It's just that we're now adding two 'last portions' which is stretching the intention somewhat.

> appendBreadcrumb is like setPostfix right? you can only append one more part?

not really - the `appendBreadcrumb` is adding another item to the array, whereas `setSuffix` is setting the final element of the array. I originally had `appendBreadcrumb` called twice, before realising that the DetailView component is re-rendering and therefore appends twice.. which is why I'm calling `setSuffix` then `appendBreadcrumb` to add another part

Reply:
another option would be to provide a formatter function to the breadcrumb component, which accepts something as a parameter and formats the final part of the breadcrumb. just thinking out loud, I have no intention to try that out.

Reply:
Or maybe `setSuffixes` so that you can set multiple suffixes if needed? Each part would be separated by a `/`.

So there would be the urlParts from the current location and then a list of additional suffix parts. A suffix part could be, for example, a <Breadcrumb> or other components.

Reply:
yes, that would work too - I wouldn't make changes now though, best to make a new issue

Reply:
https://github.com/openmsupply/open-msupply/issues/2118